### PR TITLE
ISSUE-328 Keep schema branch name unique within a schema metadata

### LINKED ISSUE #328: Schema branch should be unique only with a schema metadata
Currently, schema branch which is created is unique across all the schema metadata, this behavior should be changed so that its unique within a schema metadata

### PR DESCRIPTION
**Pull request by raju-saravanan**
ISSUE 328: Keep schema branch name unique within a schema metadata.

Closes #328

### CODE REVIEW

**Comment by raju-saravanan** [comment]
@satishd : I have rebase this PR with master, can you review this ?

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/state/SchemaVersionLifecycleStates.java [hunk 1]
```diff
@@ -15,6 +15,7 @@
  */
 package com.hortonworks.registries.schemaregistry.state;
 
+import com.fasterxml.jackson.databind.ObjectMapper;
```
Comment (satishd) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/state/SchemaVersionLifecycleStates.java]: nit: unused imports, pl do optimize imports for this class.

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/state/details/InitializedStateDetails.java [hunk 2]
```diff
@@ -25,50 +25,51 @@
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class InitializedStateDetails implements AbstractStateDetails {
 
-    public class MergedBranchDetails {
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    public static class MergeInfo {
```
Comment (satishd) [inline schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/state/details/InitializedStateDetails.java]: This is not really part of `InitializedStateDetails`, better to move out from this class.

File: schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/DefaultSchemaRegistry.java [hunk 3]
```diff
@@ -533,7 +528,29 @@ public SchemaIdVersion mergeSchemaVersion(Long schemaVersionId, SchemaVersionMer
         Collection<AggregatedSchemaBranch> aggregatedSchemaBranches = new ArrayList<>();
         for (SchemaBranch schemaBranch : getSchemaBranches(schemaName)) {
             Long rootVersion = schemaBranch.getName().equals(SchemaBranch.MASTER_BRANCH) ? null: schemaVersionLifecycleManager.getRootVersion(schemaBranch).getId();
-            aggregatedSchemaBranches.add(new AggregatedSchemaBranch(schemaBranch, rootVersion, getAllVersions(schemaBranch.getName(), schemaName)));
+            Collection<SchemaVersionInfo> schemaVersionInfos = getAllVersions(schemaBranch.getName(), schemaName);
+            schemaVersionInfos.stream().forEach(schemaVersionInfo -> {
+                SchemaVersionLifecycleContext context = null;
+                try {
+                    context = schemaVersionLifecycleManager.createSchemaVersionLifeCycleContext(schemaVersionInfo.getId(), SchemaVersionLifecycleStates.INITIATED);
```
Comment (satishd) [inline schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/DefaultSchemaRegistry.java]: Effectively, there will be a DB call for each schema version as `schemaVersionLifecycleManager.createSchemaVersionLifeCycleContext` internally needs to make that DB call.

File: schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/SchemaVersionLifecycleManager.java [hunk 4]
```diff
@@ -675,6 +685,34 @@ public SchemaIdVersion mergeSchemaVersion(Long schemaVersionId, SchemaVersionMer
         return new ImmutablePair<>(context, schemaVersionLifecycleState);
     }
 
+    public SchemaVersionLifecycleContext createSchemaVersionLifeCycleContext(Long schemaVersionId, SchemaVersionLifecycleState schemaVersionLifecycleState) throws SchemaNotFoundException {
+        // get the current state from storage for the given versionID
```
Comment (satishd) [inline schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/SchemaVersionLifecycleManager.java]: Can we file github issue for this to enhance storage manager? We already support `SearchQuery` with OrderBy, which can be enhanced to take LIMIT.

File: schema-registry/rest-service/src/test/java/com/hortonworks/registries/schemaregistry/avro/SchemaBranchLifeCycleTest.java [hunk 5]
```diff
@@ -340,14 +345,18 @@ public void deleteInvalidSchemaBranch() throws IOException, SchemaBranchNotFound
 
     @Test (expected = InvalidSchemaBranchDeletionException.class)
     public void deleteMasterBranch() throws IOException, SchemaBranchNotFoundException, InvalidSchemaException, SchemaNotFoundException, IncompatibleSchemaException, SchemaBranchAlreadyExistsException, InvalidSchemaBranchDeletionException {
+        SchemaMetadata schemaMetadata = createSchemaMetadata(testNameRule.getMethodName(), SchemaCompatibility.NONE);
```
Comment (satishd) [inline schema-registry/rest-service/src/test/java/com/hortonworks/registries/schemaregistry/avro/SchemaBranchLifeCycleTest.java]: Do you really need to add this here?

File: schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/state/SchemaVersionLifecycleStates.java (hunk 1 above)
Comment (raju-saravanan) [reply schema-registry/common/src/main/java/com/hortonworks/registries/schemaregistry/state/SchemaVersionLifecycleStates.java]: Done

**Comment by raju-saravanan** [comment]
@satishd : Addressed your PR comments, Can you please have another look ?

File: schema-registry/rest-service/src/test/java/com/hortonworks/registries/schemaregistry/avro/SchemaBranchLifeCycleTest.java (hunk 5 above)
Comment (satishd) [reply schema-registry/rest-service/src/test/java/com/hortonworks/registries/schemaregistry/avro/SchemaBranchLifeCycleTest.java]: @raju-saravanan clarified that it is needed as the branch created when a schema metadata is created for local registry server used in the test.

**Pull request closed without merging**